src/iface/ip.rs

```diff
@@ -0,0 +1,1030 @@
+                } else {
+                    // Group-specific query
+                    if self.state.has_multicast_group(group_addr) && ipv4_repr.dst_addr == group_addr {
+                        // Don't respond immediately
+                        let timeout = max_resp_time / 4;
+                        self.state.igmp_report_state = IgmpReportState::ToSpecificQuery {
+                            version, timeout: timestamp + timeout, group: group_addr
+                        };
+                    }
+                }
+            },
+            // Ignore membership reports
+            IgmpRepr::MembershipReport { .. } => (),
+            // Ignore hosts leaving groups
+            IgmpRepr::LeaveGroup{ .. } => (),
+        }
+
+        Ok(None)
+    }
+
+    #[cfg(feature = "proto-ipv6")]
+    fn process_icmpv6<'frame>(&mut self, lower: &mut impl LowerProcessor, _sockets: &mut SocketSet, timestamp: Instant,
+                              ip_repr: IpRepr, ip_payload: &'frame [u8]) -> Result<Option<Packet<'frame>>>
+    {
+        let icmp_packet = Icmpv6Packet::new_checked(ip_payload)?;
+        let checksum_caps = self.state.device_capabilities.checksum.clone();
+        let icmp_repr = Icmpv6Repr::parse(&ip_repr.src_addr(), &ip_repr.dst_addr(),
+                                          &icmp_packet, &checksum_caps)?;
+
+        #[cfg(feature = "socket-icmp")]
+        let mut handled_by_icmp_socket = false;
+
+        #[cfg(all(feature = "socket-icmp", feature = "proto-ipv6"))]
+        for mut icmp_socket in _sockets.iter_mut().filter_map(IcmpSocket::downcast) {
+            if !icmp_socket.accepts(&ip_repr, &icmp_repr.into(), &checksum_caps) { continue }
+
+            match icmp_socket.process(&ip_repr, &icmp_repr.into(), &checksum_caps) {
+                // The packet is valid and handled by socket.
+                Ok(()) => handled_by_icmp_socket = true,
+                // The socket buffer is full.
+                Err(Error::Exhausted) => (),
+                // ICMP sockets don't validate the packets in any way.
+                Err(_) => unreachable!(),
+            }
+        }
+
+        match icmp_repr {
+            // Respond to echo requests.
+            Icmpv6Repr::EchoRequest { ident, seq_no, data } => {
+                match ip_repr {
+                    IpRepr::Ipv6(ipv6_repr) => {
+                        let icmp_reply_repr = Icmpv6Repr::EchoReply {
+                            ident:  ident,
+                            seq_no: seq_no,
+                            data:   data
+                        };
+                        Ok(self.icmpv6_reply(ipv6_repr, icmp_reply_repr))
+                    },
+                    _ => Err(Error::Unrecognized),
+                }
+            }
+
+            // Ignore any echo replies.
+            Icmpv6Repr::EchoReply { .. } => Ok(None),
+
+            // Forward any NDISC packets to the ndisc packet handler
+            #[cfg(feature = "ethernet")]
```

Comment:
Is this really only needed for `ethernet`?

Reply:
I also wanna know this. Had problems with this function on my code changes